### PR TITLE
Replaces client.userId with user

### PR DESCRIPTION
**Pull request by michaelkantor**

### CODE REVIEW

File: src/client-authenticator.js [hunk 1]
```diff
@@ -208,16 +230,32 @@ class ClientAuthenticator extends Root {
...
     if (this.isTrustedDevice && userId) {
       this._restoreLastSession(userId);
+      user = this._restoreLastUser();
+      if (user) this.user = user;
     }
-    this.userId = userId;
-    if (this.sessionToken) {
+
+    if (!this.user) {
+      this.user = new UserIdentity({
```
Comment (michaelkantor) [inline src/client-authenticator.js]: The fact that a side effect of new UserIdentity({}) is to register it with the Client (not the ClientAuthenticator) is a little messy.

File: src/client-authenticator.js [hunk 2]
```diff
@@ -208,16 +230,32 @@ class ClientAuthenticator extends Root {
    * @returns {layer.ClientAuthenticator} this
    */
   connect(userId) {
+    if (!userId) userId = '';
```
Comment (nilgradisnik) [inline src/client-authenticator.js]: You can just use default parameter here `connect(userId = '')`

File: src/client-authenticator.js [hunk 3]
```diff
@@ -463,17 +534,52 @@ class ClientAuthenticator extends Root {
...
+            const sessionData = JSON.parse(global.localStorage[LOCALSTORAGE_KEYS.SESSIONDATA + this.appId]);
+            sessionData.user = DbManager.prototype._getIdentityData([this.user])[0];
+            global.localStorage[LOCALSTORAGE_KEYS.SESSIONDATA + this.appId] = JSON.stringify(sessionData);
+          } catch (e) {
+            // no-op
+          }
+        }
+        this._clientReady();
+      })
+      .once('identities:loaded-error', () => this._clientReady());
```
Comment (nilgradisnik) [inline src/client-authenticator.js]: if no displayName, set it to "you"; via property that can be customized

File: src/identity.js [hunk 4]
```diff
@@ -174,6 +174,42 @@ class UserIdentity extends Identity {
...
+ * @param {string} userId
+ */
+  _setUserId(userId) {
+    const client = this.getClient();
+    if (client) client._removeIdentity(this);
+    this.__userId = userId;
+    const encoded = encodeURIComponent(userId);
+    this.id = UserIdentity.prefixUUID + encoded;
+    this.url = `${this.getClient().url}/identities/${encoded}`;
+    if (client) client._addIdentity(this);
```
Comment (nilgradisnik) [inline src/identity.js]: Looks like you're checking for `client` in two different places and then calling `this.getClient().url` directly. You should probably just bail if you don't have a client at the beginning.

File: src/identity.js [hunk 5]
```diff
@@ -186,12 +186,13 @@ class UserIdentity extends Identity {
  */
   _setUserId(userId) {
     const client = this.getClient();
-    if (client) client._removeIdentity(this);
+    if (!client) throw new Error(LayerError.dictionary.clientMissing);
+    client._removeIdentity(this);
     this.__userId = userId;
     const encoded = encodeURIComponent(userId);
     this.id = UserIdentity.prefixUUID + encoded;
     this.url = `${this.getClient().url}/identities/${encoded}`;
```
Comment (nilgradisnik) [inline src/identity.js]: `this.getClient().url` can be just `client.url`

**Pull request merged**